design/Extend-VolumePolicies-to-support-more-actions.md

```diff
@@ -0,0 +1,205 @@
+	ActionMatchingVols := []string{}
+	NonActionMatchingVols := []string{}
+	for _, vol := range pod.Spec.Volumes {
+		action, err := volumePolicy.GetMatchAction(vol)
+		if err != nil {
+			return nil, nil, err
+		}
+		// Now if the matched action is `fs-backup` then add that Volume to the fsBackupVolumeList
+		if action != nil && action.Type == resourcepolicies.FSBackup {
+			ActionMatchingVols = append(ActionMatchingVols, vol.Name)
+		} else {
+			NonActionMatchingVols = append(NonActionMatchingVols, vol.Name)
+		}
+	}
+
+	return ActionMatchingVols, NonActionMatchingVols, nil
+}
+
+```
+
+- Making sure that `snapshot` action is skipped for PVs that do not fit the volume policy criteria:
+```go
+    // Here check if the pv fits a volume policy criteria and see if the associated action is snapshot
+	// if it is not snapshot then skip the code path for snapshotting the PV
+	volumePolicy := ib.backupRequest.ResPolicies
+	action, err := volumePolicy.GetMatchAction(obj)
+	if err != nil {
+		return err
+	}
+
+	if !(action != nil && action.Type == resourcepolicies.Snapshot) {
+		log.Info(fmt.Sprintf("skipping volume snapshot for PV %s as it does not fit the volume policy criteria for snapshot action", pv.Name))
+		ib.trackSkippedPV(obj, kuberesource.PersistentVolumes, volumeSnapshotApproach, "does not satisfy the criteria for volume policy based snapshot action", log)
+		return nil
+	}
+```
+
+- Making sure that `snapshot` action is skipped for PVCs that do not fit the volume policy criteria:
+```go
+       // check if the obj is pvc and action is a csi plugin, and if volumepolicy is specified it should satisfy the volumepolicy specified by the user else skip the execute csi action
+		if groupResource == kuberesource.PersistentVolumeClaims && actionName == csiBIAPluginName {
+			volumePolicy := ib.backupRequest.ResPolicies
+			action, err := volumePolicy.GetMatchAction(obj)
+			if err != nil {
+				return nil, itemFiles, errors.WithStack(err)
+			}
+			if !(action != nil && action.Type == resourcepolicies.Snapshot) {
+				log.Info(fmt.Sprintf("skipping csi volume snapshot for PVC %s as it does not fit the volume policy criteria for snapshot action", namespace+" /"+name))
+				ib.trackSkippedPV(obj, kuberesource.PersistentVolumeClaims, volumeSnapshotApproach, "does not satisfy the criteria for volume policy based snapshot action", log)
+				continue
+			}
+		}
+```
+
+## Future Implementation
+It makes sense to add more specific actions in the future, once we deprecate the legacy opt-in/opt-out approach to keep things simple. Another point of note is, csi related action can be 
+easier to implement once we decide to merge the csi plugin into main velero code flow.
+In the future, we envision the following actions that can be implemented:
+- `snapshot-native`: only use volume snapshotter (native cloud provider snapshots), do nothing if not present/not compatible
```

Comment:
This is a little trivial b/c it's about future Impl, but IMO these 3 actions should be merged into the existing `snapshot` action and controlled by the `Parameters` of the action.

Reply:
@reasonerjt Added a note regarding this.